packages/nx-python/src/executors/build/resolvers/project.ts

```diff
@@ -99,6 +133,23 @@ export class ProjectDependencyResolver {
           continue;
         } else {
           dep.version = depPyproject.tool.poetry.version;
+
+          if (targetOptions?.customSourceUrl) {
+            const [newSources, newSourceName] = resolveDuplicateSources(
+              buildTomlData.tool.poetry.source,
+              {
+                name: targetOptions.customSourceName,
+                url: targetOptions.customSourceUrl,
+              }
+            );
+            if (newSourceName != targetOptions.customSourceName) {
+              this.logger.info(
+                chalk`  Duplicate source for {blue.bold ${dep.name}} renamed to ${newSourceName}`
+              );
+            }
+            buildTomlData.tool.poetry.source = newSources;
+            dep.source = newSourceName;
+          }
```

Comment:
I don't think that we should use `this.options`. Using `this.options` would apply the source for the parent to all its dependencies. I have instead used the child dependencies configured build options so that you can specify different children from different sources. This may be more clear if you look at the test I have added.

Reply:
The reason I have created the resolve dependency options is to cater for the use case when multiple dependencies are sourced from different repositories but have been configured with the same name..

Reply:
hmm.. got it, the only problem with that is you can't use default values in the `nx.json` or use CLI options `--customSourceName something`.

perhaps that is not a problem, if the dependency project does not have those options is going to compile without the source.

and for this scenario makes sense to generate a unique source name, maybe instead of using an md5 function you can concat with the project name `${depName}-${customSourceName}` that will make SonarCloud happy, and also avoid a crypto function.

What do you think?

Reply:
Yeah I suppose that makes sense, shouldn't ever have a duplicate for the given dependency. I'll update it now.

Reply:
Cool, many thanks, SonarCloud also complained about the complexity of the function, could you also refactor the code block to a different dedicated function?

Look: https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=41&id=lucasvieirasilva_nx-plugins&open=AYYttJkNTlC5F8A37_w7

Reply:
Actually looking at implementing that it looks a bit gross. You end up with a source defined for every dependency in a package. Even if the URLs match. What I have implemented already instead attempts to add a source with the name defined, if the name matches it will use a hash of the URL on the end. This means that even if other packages use the same name / url combination they don't keep adding additional sources to the list.

Reply:
hmm.. perhaps we should never use the `customSourceName` and always generate the source name based on the URL, so you just need to check if the hash exists, if no, you don't add it to the source object.

I'm happy to flag it as safe in SonarCloud.

Reply:
That's an option, makes it a bit less obvious what the source is for but wouldn't be hard to lookup. 🤷 Happy to go with whatever you prefer.

Reply:
you're right, I also noticed you added the name + hash, which makes it more readable, I've flagged the security issue as safe on SonarCloud, could just have a look at the code smell?

Reply:
Thanks. 

What do you mean by the code smell is this context sorry? Is there a specific thing you want me to change.

Reply:
@niall-mediasuite This one https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&types=CODE_SMELL&pullRequest=41&id=lucasvieirasilva_nx-plugins&open=AYYttJkNTlC5F8A37_w7

When the function is getting big SonarCloud recommends to refactor some of the logic to a dedicated function.

Reply:
Ah thanks, taking a look now.

Reply:
I think I've fixed it now... Not really sure how to tell.

Reply:
@niall-mediasuite Almost, your last change reduced the complexity but added another code smell.

https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&sinceLeakPeriod=true&pullRequest=41&id=lucasvieirasilva_nx-plugins

You can see the SonarCloud report on the GitHub pull request check tab.

Reply:
Ok, should finally be fixed now. Yeah sorry I see it there but it said that all checks passed so I assume it was ok. :think

Reply:
Ok, should finally be fixed now. Yeah sorry I see it there but it said that all checks passed so I assume it was ok. 🤔

Reply:
Never mind, still needs work.

Reply:
Looks like its fixed now as far as I can tell.

Reply:
Yep, looks good now, many thanks!